File: typescript/utils/logger.ts
export type LogLevel = 'debug' | 'info' | 'warn' | 'error' | 'silent';

const LEVELS: Record<LogLevel, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
  silent: 4,
};

export class Logger {
  private level: number;

  constructor(level: LogLevel = 'info') {
    this.level = LEVELS[level];
  }

  setLevel(level: LogLevel): void {
    this.level = LEVELS[level];
  }

  debug(msg: string, data?: Record<string, unknown>): void {
    if (this.level <= LEVELS.debug) this.log('DEBUG', msg, data);
  }

  info(msg: string, data?: Record<string, unknown>): void {
    if (this.level <= LEVELS.info) this.log('INFO', msg, data);
  }

  warn(msg: string, data?: Record<string, unknown>): void {
    if (this.level <= LEVELS.warn) this.log('WARN', msg, data);
  }

  error(msg: string, data?: Record<string, unknown>): void {
    if (this.level <= LEVELS.error) this.log('ERROR', msg, data);
  }

  private log(level: string, msg: string, data?: Record<string, unknown>): void {
    const ts = new Date().toISOString();
    const extra = data ? ' ' + JSON.stringify(data) : '';
    const line = `[${ts}] ${level} spider-browser: ${msg}${extra}`;
    if (level === 'ERROR') {
      console.error(line);
    } else if (level === 'WARN') {
      console.warn(line);
    } else {
      console.log(line);
    }
  }
}

/** Shared default logger instance. */
export const logger = new Logger('info');

File: go/logger.go
package spiderbrowser

import (
	"fmt"
	"os"
	"time"
)

// LogLevel controls logging verbosity.
type LogLevel int

const (
	LogDebug LogLevel = iota
	LogInfo
	LogWarn
	LogError
	LogSilent
)

// Logger provides structured logging for spider-browser.
type Logger struct {
	level LogLevel
}

var defaultLogger = &Logger{level: LogInfo}

// SetLogLevel sets the global log level.
func SetLogLevel(level LogLevel) {
	defaultLogger.level = level
}

func (l *Logger) debug(msg string, data ...any) {
	if l.level <= LogDebug {
		l.log("DEBUG", msg, data...)
	}
}

func (l *Logger) info(msg string, data ...any) {
	if l.level <= LogInfo {
		l.log("INFO", msg, data...)
	}
}

func (l *Logger) warn(msg string, data ...any) {
	if l.level <= LogWarn {
		l.log("WARN", msg, data...)
	}
}

func (l *Logger) logError(msg string, data ...any) {
	if l.level <= LogError {
		l.log("ERROR", msg, data...)
	}
}

func (l *Logger) log(level, msg string, data ...any) {
	ts := time.Now().UTC().Format(time.RFC3339Nano)
	extra := ""
	if len(data) > 0 {
		extra = fmt.Sprintf(" %v", data)
	}
	line := fmt.Sprintf("[%s] %s spider-browser: %s%s\n", ts, level, msg, extra)
	if level == "ERROR" || level == "WARN" {
		fmt.Fprint(os.Stderr, line)
	} else {
		fmt.Fprint(os.Stdout, line)
	}
}
